File: HW/HW 12/Solution.py
from typing import List
class Solution:
    def minCostToSupplyWater(self, n:int, wells: List[int], pipes: List[List[int]]) -> int:
        edges = []
        for i in range(len(wells)):
            edges.append([0, i+1, wells[i]])

        for pipe in pipes:
            edges.append(pipe)
        
        edges.sort(key = lambda x: x[2])
        parent = [i for i in range(n+1)]

        cost = 0
        for edge in edges:
            u, v, weight = edge
            parentU = self.find(u,parent)
            parentV = self.find(v,parent)
            if parentU != parentV:
                cost += weight
                parent[parentU] = parentV
        
        return cost
    
    def find(self, x:int, parent: List[int]) -> int:
        if parent [x] != x:
            parent[x] = self.find(parent[x], parent)
        return parent[x]
    
solution = Solution()
# Test Case 1
n1 = 2
wells1 = [1, 1]
pipes1 = [[1, 2, 1], [1, 2, 2]]
output1 = solution.minCostToSupplyWater(n1, wells1, pipes1)
print("Test Case 1:", output1)  # should be 2

# Test Case 2
n2 = 3
wells2 = [1, 2, 2]
pipes2 = [[1, 2, 1], [2, 3, 1]]
output2 = solution.minCostToSupplyWater(n2, wells2, pipes2)
print("Test Case 2:", output2)  # should be 3

File: HW/HW 12/Solution.java
import java.util.*;
public class Solution {
    public int minCostToSupplyWater(int n, int [] wells, int[][] pipes){
        List<int[]> edges = new ArrayList<>();
        for(int i = 0;i<wells.length;i++){
            edges.add(new int[]{0,i+1,wells[i]});
        }
        for(int[] pipe:pipes){
            edges.add(pipe);
        }
        Collections.sort(edges,(a,b) -> a[2] - b[2]);
        int[] parent = new int[n+1];
        for(int i = 0;i<=n;i++){
            parent[i]=i;
        }
        int cost = 0;
        for(int[] edge:edges){
            int u = edge[0];
            int v = edge[1];
            int weight = edge[2];
            int parentU = find(u, parent);
            int parentV = find(v, parent);
            if(parentU != parentV){
                cost += weight;
                parent[parentU] = parentV;
            }
        }
        return cost;
    }
    private int find(int x, int[] parent){
        if(parent[x] != x) parent[x] = find(parent[x], parent);
        return parent[x];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test Case 1
        int n1 = 2;
        int[] wells1 = {1, 1};
        int[][] pipes1 = {{1, 2, 1}, {1, 2, 2}};
        int output1 = solution.minCostToSupplyWater(n1, wells1, pipes1);
        System.out.println("Test Case 1: " + (output1)); //should be 2

        // Test Case 2
        int n2 = 3;
        int[] wells2 = {1, 2, 2};
        int[][] pipes2 = {{1, 2, 1}, {2, 3, 1}};
        int output2 = solution.minCostToSupplyWater(n2, wells2, pipes2);
        System.out.println("Test Case 2: " + (output2));//should be 3
    }
}